src/code_42e0_len_1f60.c

```diff
@@ -1,27 +1,120 @@
+};
+
+s32 gRandSeed = 1;
+
+f32 D_80074414[] = {
+    1.0f, 1.273187f, 1.27303f, 1.272768f, 1.272402f, 1.271932f, 1.271358f, 1.270681f, 1.269902f, 1.269021f, 1.268038f,
+    1.266956f, 1.265774f, 1.264494f, 1.263116f, 1.261643f, 1.260075f, 1.258413f, 1.256659f, 1.254815f, 1.252881f,
+    1.250859f, 1.248752f, 1.24656f, 1.244285f, 1.241929f, 1.239494f, 1.236981f, 1.234393f, 1.231731f, 1.228997f,
+    1.226192f, 1.22332f, 1.220382f, 1.217379f, 1.214315f, 1.211189f, 1.208006f, 1.204766f, 1.201471f, 1.198124f,
+    1.194727f, 1.191281f, 1.187787f, 1.18425f, 1.180669f, 1.177047f, 1.173386f, 1.169687f, 1.165952f, 1.162184f,
+    1.158384f, 1.154553f, 1.150693f, 1.146806f, 1.142893f, 1.138957f, 1.134998f, 1.131018f, 1.127019f, 1.123002f,
+    1.118969f, 1.11492f, 1.110858f, 1.106783f, 1.102697f, 1.098601f, 1.094496f, 1.090384f, 1.086266f, 1.082142f,
+    1.078014f, 1.073883f, 1.06975f, 1.065616f, 1.061482f, 1.057348f, 1.053216f, 1.049087f, 1.04496f, 1.040838f,
+    1.036721f, 1.03261f, 1.028504f, 1.024406f, 1.020316f, 1.016234f, 1.01216f, 1.008097f, 1.004043f, 1.0f
+};
+
+s32 D_80074580[] = {
+    0xE7000000, 0x00000000, 0xD7000000, 0xFFFFFFFF, 0xE3000A01, 0x00000000, 0xE3000C00, 0x00000000, 0xE3000D01,
+    0x00000000, 0xE3000F00, 0x00000000, 0xE3001201, 0x00000000, 0xE3001001, 0x00000000, 0xE3001402, 0x00000C00,
+    0xDF000000, 0x00000000
+};
 
 void sin_cos_rad(f32 rad, f32* outSinTheta, f32* outCosTheta);
 void func_80029860(s32 romStart, s32 vramDest, s32 length);
 
 #define ROM_CHUNK_SIZE 0x2000
 
 void poll_rumble(void) {
-    // TODO: replace with defines
     nuContRmbCheck(0);
     nuContRmbModeSet(0, 2);
 }
 
-INCLUDE_ASM(s32, "code_42e0_len_1f60", start_rumble);
+void start_rumble(s32 freq, s32 frame) {
+    if (GAME_STATUS->demoState == 0) {
+        u16* sym = &D_80074260;
 
-INCLUDE_ASM(s32, "code_42e0_len_1f60", func_80028F8C);
+        if (*sym != 0) {
+            s32 symx2 = *sym * 2;
 
-INCLUDE_ASM(s32, "code_42e0_len_1f60", func_80028FE0);
+            if (frame > symx2) {
+                frame = symx2;
+            }
+
+            if (nuContRmbCheck(0) == 0) {
+                nuContRmbModeSet(0, 2);
+                nuContRmbStart(0, freq, frame);
+            }
+        }
+    }
+}
+
+void func_80028F8C(void) {
+    s32* sym = &D_80074264;
+    u16* sym2;
+
+    if (*sym != GAME_STATUS->currentButtons) {
+        *sym = GAME_STATUS->currentButtons;
+        func_80028FE0();
+    }
+
+    sym2 = &D_80074260;
+    if (*sym2 != 0) {
+        (*sym2)--;
+    }
+}
+
+void func_80028FE0(void) {
+    D_80074260 = 300;
+}
 
 f32 length2D(f32 x, f32 y) {
     return sqrtf(SQ(x) + SQ(y));
 }
 
-INCLUDE_ASM(HeapNode*, "code_42e0_len_1f60", _heap_create, void* addr, s32 size);
+HeapNode* _heap_create(s32* addr, u32 size) {
+    if (size < 32) {
+        return (HeapNode*) - 1;
+    } else {
+        HeapNode* heapNode = ALIGN16((s32)addr);
+
+        size -= ((char*)heapNode - (char*)addr);
```

Comment:
I'd prefer u8/s8 for portability reasons here. (Char is not guaranteed to be 1 byte long.)